lib/src/stream_closer.dart

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'async_memoizer.dart';
+
+/// A [StreamTransformer] that allows the caller to forcibly close the
+/// transformed [Stream](s).
+///
+/// When [close] is called, any stream (or streams) transformed by this
+/// transformer that haven't already completed or been cancelled will emit a
+/// done event and cancel their underlying subscriptions.
+///
+/// Note that unlike most [StreamTransformer]s, each instance of [StreamCloser]
+/// has its own state (whether or not it's been closed), so it's a good idea to
+/// construct a new one for each use unless you need to close multiple streams
+/// at the same time.
+class StreamCloser<T> extends StreamTransformerBase<T, T> {
+  /// Whether [close] has been called.
+  bool get isClosed => _closeMemo.hasRun;
+
+  /// The subscriptions to streams passed to [bind].
+  final _subscriptions = <StreamSubscription<T>>{};
+
+  /// The controllers for streams returned by [bind].
+  final _controllers = <StreamController<T>>{};
+
+  /// Closes all transformed streams.
+  ///
+  /// Returns a future that completes when all inner subscriptions'
+  /// [StreamSubscription.cancel] futures have completed. Note that a stream's
+  /// subscription won't be canceled until the transformed stream has a
+  /// listener.
+  ///
+  /// If a transformed stream is listened to after [close] is called, the
+  /// original stream will be listened to and then the subscription immediately
+  /// canceled. If that cancellation throws an error, it will be silently
+  /// ignored.
+  Future<void> close() => _closeMemo.runOnce(() {
```

Comment:
Is `AsyncMemoizer` worth the overhead here. It has an extra completer with its future, and a call to `Future.sync`. That's two extra futures and a completer for no real benefit.
Couldn't it just be:
```dart
Future<void>? _closeFuture;
bool get isClosed => _closeFuture != null;
Future<void> close() => _closeFuture ??= _close();
Future<void> _close() {
  // code here.
}
```
(or inline `_close` in `close` as `() { ... } ();`).